ocaml-lsp-server/src/custom_requests/req_type_enclosing.ml

```diff
@@ -0,0 +1,191 @@
+open Import
+
+let capability = ("handleTypeEnclosing", `Bool true)
+
+let meth = "ocamllsp/typeEnclosing"
+
+type params =
+  { text_document_position : Lsp.Types.TextDocumentPositionParams.t
+  ; index : int
+  ; range_end : Position.t option
+  ; verbosity : int
+  }
+
+let expected_params =
+  `Assoc
+    [ ("index", `String "uinteger")
+    ; ("rangeEnd?", `String "<Position>")
+    ; ("verbosity?", `String "uinteger")
+    ; ("position", `String "<Position>")
+    ; ("textDocument", `String "<TextDocumentIdentifier>")
+    ]
+
+let index_of_yojson params =
+  match List.assoc_opt "index" params with
+  | Some (`Int index) -> Some index
+  | _ ->
+    (* If the parameter is incorrectly formatted or missing, we refuse to build
+       the parameter, [index] is mandatory. *)
+    None
+
+let verbosity_of_yojson params =
+  match List.assoc_opt "verbosity" params with
+  | Some (`Int verbosity) -> verbosity
+  | _ ->
+    (* If the parameter is incorrectly formatted or missing, it is assumed that
+       the we ask for a verbosity level set to 0. *)
+    0
+
+let range_end_of_yojson params =
+  match List.assoc_opt "rangeEnd" params with
+  | Some range_end -> (
+    try Some (Position.t_of_yojson range_end) with _ -> None)
+  | _ ->
+    (* If the parameter is incorrectly formatted or missing, it is assumed that
+       the we do not provide rangeEnd parameter. *)
+    None
+
+let raise_invalid_params ?data ~message () =
+  let open Jsonrpc.Response.Error in
+  raise @@ make ?data ~code:Code.InvalidParams ~message ()
+
+let of_yojson = function
+  | `Assoc params as full_params ->
+    let verbosity = verbosity_of_yojson params in
+    let range_end = range_end_of_yojson params in
+    let open Option.O in
+    let* index = index_of_yojson params in
+    let text_document_position =
+      Lsp.Types.TextDocumentPositionParams.t_of_yojson full_params
+    in
+    Some { index; range_end; verbosity; text_document_position }
+  | _ -> None
+
+let of_yojson_exn = function
+  | None -> raise_invalid_params ~message:"Expected params but received none" ()
+  | Some params -> (
+    match of_yojson params with
+    | Some params -> params
+    | None ->
+      let data =
+        `Assoc
+          [ ("expectedParams", expected_params)
+          ; ("receivedParams", (params :> Json.t))
+          ]
+      in
+      raise_invalid_params ~data ~message:"Unexpected params format" ())
```

Comment:
All of this error handling is done at the wrong layer. If we get a json encoding error, there should be a single place that converts into the proper error code. So instead of collecting the option type and worrying about a pretty error message, doing the easy thing and just raising the same exception that ppx_yojson_conv raises is the correct thing to do. If we want to improve error messages for json, we should just do it in one place for all requests and rely on the exceptions being informative.